src/framing/Frame.cpp

```diff
@@ -94,78 +94,120 @@ std::ostream& operator<<(std::ostream& os, ErrorCode errorCode) {
+
+std::string to_string(FrameFlags frameFlags,
+                      const std::vector<std::pair<FrameFlags, std::string>>& allowedFlags)
+{
+  FrameFlags foundFlags = FrameFlags::EMPTY;
 
   std::stringstream ss;
   std::string delimeter = "";
-  if (!!(frameFlags & FrameFlags::NEXT)) {
-    ss << "NEXT";
-    delimeter = "|";
-  }
-  if (!!(frameFlags & FrameFlags::COMPLETE)) {
-    ss << delimeter << "COMPLETE";
-    delimeter = "|";
-  }
-  if (!!(frameFlags & FrameFlags::FOLLOWS)) {
-    ss << delimeter << "FOLLOWS";
-    delimeter = "|";
+  for (auto& pair : allowedFlags) {
+    if (!!(frameFlags & pair.first)) {
+      ss << delimeter << pair.second;
+      delimeter = "|";
+      foundFlags |= pair.first;
+    }
   }
-  if (!!(frameFlags & FrameFlags::METADATA)) {
-    ss << delimeter << "METADATA";
-    delimeter = "|";
-  }
-  if (!!(frameFlags & FrameFlags::IGNORE)) {
-    ss << delimeter << "IGNORE";
-    delimeter = "|";
-  }
-  if (!delimeter.empty()) {
-    return os << ss.str();
+
+  if (foundFlags != frameFlags) {
+    ss <<frameFlags;
+  } else if (delimeter.empty()) {
+    ss << sFrameFlags_EMPTY;
   }
-  return os << "EMPTY";
+  return ss.str();
+}
+
+std::ostream& operator<<(std::ostream& os, FrameFlags frameFlags) {
+  std::bitset<16> flags(static_cast<uint16_t>(frameFlags));
+  return os << flags;
 }
 
 std::ostream& operator<<(std::ostream& os, const FrameHeader& header) {
   return os << header.type_ << "[" << header.flags_ << ", " << header.streamId_
             << "]";
 }
+
+std::ostream& operator<<(std::ostream& os,
+    const std::pair< const FrameHeader&,
+        const std::vector<std::pair<FrameFlags, std::string>>&
+      > & headerAndAllowedFlags
+  ) {
+  const auto& header = headerAndAllowedFlags.first;
+  const auto& allowedFlags = headerAndAllowedFlags.second;
+  return os << header.type_ << "[" << to_string(header.flags_, allowedFlags)
+            << ", " << header.streamId_ << "]";
+}
+
 /// @}
 
 std::ostream& operator<<(std::ostream& os, const Frame_REQUEST_Base& frame) {
-  return os << frame.header_ << "(" << frame.requestN_ << ", "
+  return os << std::make_pair(frame.header_,
+                              std::vector<std::pair<FrameFlags, std::string>> {})
+            << "("
+            << frame.requestN_ << ", "
             << frame.payload_;
 }
 
 std::ostream& operator<<(std::ostream& os, const Frame_REQUEST_N& frame) {
-  return os << frame.header_ << "(" << frame.requestN_ << ")";
+  return os << std::make_pair(frame.header_,
+                              std::vector<std::pair<FrameFlags, std::string>> {})
+            << "(" << frame.requestN_ << ")";
 }
 
 std::ostream& operator<<(
     std::ostream& os,
     const Frame_REQUEST_RESPONSE& frame) {
-  return os << frame.header_ << ", " << frame.payload_;
+  static const std::vector<std::pair<FrameFlags, std::string>> allowedFlags {
+      {FrameFlags::METADATA, sFrameFlags_METADATA},
+      {FrameFlags::FOLLOWS, sFrameFlags_FOLLOWS} };
```

Comment:
Any reason why we are not using the AllowedFlags member variable of the Frame struct (instead of redefining it here) ? Then you could write a simple utility function which takes a list of FrameFlags and returns a string (or vector of strings).

```
std::ostream& operator<<(std::ostream& os, const Frame_REQUEST_RESPONSE& frame) {
     return os << frame.header_ << ", " << frame.payload_ << toString(frame.AllowedFlags);
}
```